src/adapter/parse-semverish.ts

```diff
@@ -0,0 +1,55 @@
+const MAJOR = 1;
+const MINOR = 2;
+const PATCH = 3;
+const PRERELEASE = 5;
+const PRERELEASE_TAG = 5;
+const PRERELEASE_VERSION = 6;
+const PATTERN_MAJOR_MINOR_PATCH = "([0-9]+).([0-9]+).([0-9]+)";
```

Comment:
We could catch all cases with a single regex:
```
/(\d+)\.(\d+)\.(\d+)(-(.*)\.(\d+))?/i
```

_Note that dots must be escaped, because it's a special regex character and that matches literally anything._

All suffixes for pre-releases follow the ordering of the alphabet by coincidence:

1. `alpha`
2. `beta`
3. `rc`

We can use that to our advantage and do a direct string comparison. If we have som sort of function with the type signature `compareSemVer(a: string, b: string): -1 | 0 | 1` (similar signature as `Array.prototype.sort()`) we could easily compare SemVer strings.

```js
const available = {
  "10.2.0": createFutureRenderer();
  "10.0.0": createRenderer(),
  "8.5.0": createLegacyRenderer(),
};

const preact = "10.0.5";
const idx = Object.keys(available).findIndex(key => compareSemVer(key, version) <= 0);
const renderer = available[idx];
```

_Note: The terminology `Renderer` is a bit weird imo. I adopted it from the `react-devtools` but I'm not much of a fan of it and didn't have a better idea._

Reply:
The dot escaping is a good catch, how could I forget about this 🤦‍♂ 

Regarding the pre-release suffixes I tried to stay on the safe side as semver seems to allow almost everything as pre-release suffix identifiers :/

With `Object.keys(available)` we need to be careful as object key order stability was not always part of the JS spec (I think it now is, but not sure). I guess object keys sorting is stable for FF and Chrome thus we should be safe.